### PR TITLE
Refactor Filesystem Code for Clarity

### PR DESCRIPTION
**Pull request by AtariDreams**
This makes the code more readable as well as lowers operations to code paths that actually need them

### CODE REVIEW

File: stl/inc/filesystem [hunk 1]
```diff
@@ -1513,11 +1513,11 @@ namespace filesystem {
             // at this point, the next element is a standard filename from relative_path(), and _Position
             // points at the preferred-separator or fallback-separator after the previous element
             while (_Is_slash(*_Position)) { // advance to the start of the following path element
-                if (_Get_unwrapped(++_Position) == _End) { // "magic" empty element selected
-                    --_Position;
+                if (_Get_unwrapped(_Position + 1) == _End) { // "magic" empty element selected
```
Comment (miscco) [inline stl/inc/filesystem:1516]: This adds needless work, as `_Position` is incremented twice

File: stl/src/filesys.cpp [hunk 2]
```diff
@@ -74,6 +69,10 @@ _FS_DLL wchar_t* __CLRCALL_PURE_OR_CDECL _Read_dir(
             return _Strcpy(_Dest, &_Dentry.cFileName[0]);
         }
     }
+
+    // fail
```
Comment (miscco) [inline stl/src/filesys.cpp:73]: That comment does not add value

File: stl/src/filesys.cpp [hunk 3]
```diff
@@ -122,18 +121,17 @@ _FS_DLL void* __CLRCALL_PURE_OR_CDECL _Open_dir(
         && (_Dentry.cFileName[1] == L'\0'
             || _Dentry.cFileName[1] == L'.' && _Dentry.cFileName[2] == L'\0')) { // skip "." and ".."
         _Read_dir(_Dest, _Handle, _Ftype);
-        if (_Dest[0] != L'\0') {
```
Comment (miscco) [inline stl/src/filesys.cpp:125]: I find those changes to be less readable

File: stl/src/filesys.cpp [hunk 4]
```diff
@@ -60,12 +60,7 @@ _FS_DLL wchar_t* __CLRCALL_PURE_OR_CDECL _Read_dir(
     wchar_t (&_Dest)[_MAX_FILESYS_NAME], void* _Handle, file_type& _Ftype) { // read a directory entry
     WIN32_FIND_DATAW _Dentry;
 
-    for (;;) {
-        if (FindNextFileW(_Handle, &_Dentry) == 0) { // fail
-            _Ftype = file_type::unknown;
-            return _Strcpy(_Dest, L"");
-        }
-
+    while (FindNextFileW(_Handle, &_Dentry)) {
```
Comment (miscco) [inline stl/src/filesys.cpp:63]: Implicit conversion to bool are conventionally avoided

File: stl/src/filesys.cpp [hunk 5]
```diff
@@ -226,22 +222,26 @@ _FS_DLL unsigned long long __CLRCALL_PURE_OR_CDECL _Hard_links(const wchar_t* _F
...
+
     FILE_STANDARD_INFO _Info = {0};
-    bool _Ok                 = false;
 
-    if (_Handle != INVALID_HANDLE_VALUE) { // get file info
-        _Ok = GetFileInformationByHandleEx(_Handle, FileStandardInfo, &_Info, sizeof(_Info)) != 0;
-        CloseHandle(_Handle);
-    }
+    // get file info
+    BOOL _Ok = GetFileInformationByHandleEx(_Handle, FileStandardInfo, &_Info, sizeof(_Info));
```
Comment (miscco) [inline stl/src/filesys.cpp:232]: I think BOOL is wrong here

File: stl/src/filesys.cpp [hunk 6]
```diff
@@ -416,16 +416,18 @@ _FS_DLL int __CLRCALL_PURE_OR_CDECL _Rename(const wchar_t* _Fname1, const wchar_
...
-        _Ok             = SetFilePointerEx(_Handle, _Large, nullptr, FILE_BEGIN) != 0 && SetEndOfFile(_Handle) != 0;
-        CloseHandle(_Handle);
+    if (_Handle == INVALID_HANDLE_VALUE) {
+        return GetLastError();
     }
+
+    // set file pointer to new size and trim
+    LARGE_INTEGER _Large;
+    _Large.QuadPart = _Newsize;
+    bool _Ok        = SetFilePointerEx(_Handle, _Large, nullptr, FILE_BEGIN) && SetEndOfFile(_Handle);
```
Comment (miscco) [inline stl/src/filesys.cpp:429]: Could be const

**Review by miscco — CHANGES_REQUESTED** [review]

**Comment by StephanTLavavej** [comment]
As a maintainer, I find this PR to be concerning. Every code change is risky to some degree (even innocent-seeming changes are occasionally risky, up to and including ABI breaks), and also costs time to review. With a brief glance at the changes, I can see that there's some good stuff here (e.g. I particularly like the use of the `SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE` constant which is now available), but I have these specific concerns:

1. This is a "grab bag" of different refactorings, which is harder to review. A PR that performs only one kind of transformation (e.g. [De Morgan](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) changes) is significantly less risky and easier to review - even when many sites are being changed. The dangerous thing about mixed changes is that we can nod our heads to a bunch of safe changes and miss the one scary/wrong change in the middle.
   + It can sometimes be acceptable to have one PR with several kinds of refactorings cleanly grouped into separate commits. I suppose I would accept separate commits here.
2. Substantial behavior changes (whether bug fixes or performance improvements) should not be mixed with stylistic refactorings - that is, #2767 should be addressed in an isolated PR. Occasionally, because we are human (and slightly lazy), it is okay to perform one tiny cleanup in a PR that's otherwise dedicated to fixing a bug or improving performance. However, any substantial mixing really must be avoided. (Same rationale as avoiding mixed refactorings, but even stronger.)
   + Putting behavior changes and refactorings into separate commits doesn't solve this concern, they really need to be separate PRs.
3. It would be helpful for your PR description to explain why each change is being made. For some changes, I can immediately see the clarity improvement (e.g. De Morganing "test existing files for equivalence" so that it reads "if the serial numbers are the same, and the high parts are the same, and the low parts are the same, then we're equivalent" avoids the confusing negation in the original form - great!). For others, I am unsure how they are improving consistency - e.g. some conditions are being reversed to return-on-success (e.g. `__std_fs_get_current_path()`), but others are being reversed to return-on-failure (e.g. `_Open_dir()`). Having a changelog would help me and other reviewers understand the changes. It doesn't need to be *super* detailed, it just needs to mention why each function is being changed.
   + This is also useful for "source code archaeology" - as the STL is immortal, we occasionally need to look up why things were changed 5 or 10 years ago. Having a PR description that explains each change will be helpful if we need to read it in 2027 or 2032 in our flying cars.

File: stl/src/filesys.cpp [hunk 7]
```diff
@@ -173,9 +170,7 @@ _FS_DLL wchar_t* __CLRCALL_PURE_OR_CDECL _Temp_get(wchar_t (&_Dest)[_MAX_FILESYS
 
 _FS_DLL int __CLRCALL_PURE_OR_CDECL _Make_dir(const wchar_t* _Fname, const wchar_t*) {
     // make a new directory (ignore attributes)
-    int _Ans = CreateDirectoryW(_Fname, nullptr);
-
-    if (_Ans != 0) {
+    if (CreateDirectoryW(_Fname, nullptr)) {
```
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:173]: `!= 0`

File: stl/src/filesys.cpp [hunk 8]
```diff
@@ -226,23 +221,27 @@ _FS_DLL unsigned long long __CLRCALL_PURE_OR_CDECL _Hard_links(const wchar_t* _F
...
-    if (_Handle != INVALID_HANDLE_VALUE) { // get file info
-        _Ok = GetFileInformationByHandleEx(_Handle, FileStandardInfo, &_Info, sizeof(_Info)) != 0;
-        CloseHandle(_Handle);
-    }
-    return _Ok ? _Info.NumberOfLinks : static_cast<unsigned long long>(-1);
+    // get file info
+    const auto _Ok = GetFileInformationByHandleEx(_Handle, FileStandardInfo, &_Info, sizeof(_Info));
+    CloseHandle(_Handle);
+    return _Ok != 0 ? _Info.NumberOfLinks : static_cast<unsigned long long>(-1);
 #else // _CRT_APP
```
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:234]: These preprocessor guards can be enhanced to `^^^ / vvv` (but possibly in a separate change, see #351)

File: stl/src/filesys.cpp [hunk 9]
```diff
@@ -351,10 +350,10 @@ _FS_DLL int __CLRCALL_PURE_OR_CDECL _Equivalent(
...
         return 0;
     } else { // test existing files for equivalence
-        return _Info1.VolumeSerialNumber != _Info2.VolumeSerialNumber
-                    || memcmp(&_Info1.FileId, &_Info2.FileId, sizeof(_Info1.FileId)) != 0
-                 ? 0
-                 : 1;
+        return _Info1.VolumeSerialNumber == _Info2.VolumeSerialNumber
+                    && memcmp(&_Info1.FileId, &_Info2.FileId, sizeof(_Info1.FileId)) == 0
+                 ? 1
+                 : 0;
```
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:356]: I don't see why we can't compare `FILE_ID_INFO` as a whole thing with a single `memcmp`

File: stl/src/filesystem.cpp [hunk 10]
```diff
@@ -224,7 +225,7 @@ void __stdcall __std_fs_close_handle(const __std_fs_file_handle _Handle) noexcep
     const HANDLE _As_plain_handle = reinterpret_cast<HANDLE>(_Handle);
 
     FILE_BASIC_INFO _Ex_info;
-    if (GetFileInformationByHandleEx(_As_plain_handle, FileBasicInfo, &_Ex_info, sizeof(_Ex_info)) != 0) {
+    if (GetFileInformationByHandleEx(_As_plain_handle, FileBasicInfo, &_Ex_info, sizeof(_Ex_info))) {
```
Comment (AlexGuteniev) [inline stl/src/filesystem.cpp:228]: keep `!= 0`

File: stl/src/filesys.cpp [hunk 11]
```diff
@@ -379,10 +375,10 @@ _FS_DLL int __CLRCALL_PURE_OR_CDECL _Equivalent(
     } else if (!_Ok1 || !_Ok2) {
         return 0;
     } else { // test existing files for equivalence
-        return _Info1.dwVolumeSerialNumber != _Info2.dwVolumeSerialNumber
-                    || _Info1.nFileIndexHigh != _Info2.nFileIndexHigh || _Info1.nFileIndexLow != _Info2.nFileIndexLow
-                 ? 0
-                 : 1;
+        return _Info1.dwVolumeSerialNumber == _Info2.dwVolumeSerialNumber
```
Comment (AlexGuteniev) [inline stl/src/filesys.cpp:378]: here as well only `memcmp` can be used

Comment (AtariDreams) [reply stl/src/filesys.cpp:378]: Not really, since we are only comparing three fields, and these are not all of them

File: stl/src/filesys.cpp [hunk 12]
```diff
@@ -221,39 +217,43 @@ _FS_DLL file_type __CLRCALL_PURE_OR_CDECL _Lstat(const wchar_t* _Fname, perms* _
...
 
-    if (_Handle != INVALID_HANDLE_VALUE) { // get file info
-        _Ok = GetFileInformationByHandleEx(_Handle, FileStandardInfo, &_Info, sizeof(_Info)) != 0;
-        CloseHandle(_Handle);
-    }
-    return _Ok ? _Info.NumberOfLinks : static_cast<unsigned long long>(-1);
+    // get file info
+    const auto _Ok = GetFileInformationByHandleEx(_Handle, FileStandardInfo, &_Info, sizeof(_Info));
+    CloseHandle(_Handle);
+    return _Ok != 0 ? _Info.NumberOfLinks : static_cast<unsigned long long>(-1);
```
Comment (StephanTLavavej) [inline stl/src/filesys.cpp:234]: This should use `uintmax_t` for consistency with the return type and `#else` codepath.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks, this looks good! I found no correctness issues and only minor stylistic issues, so I'll go ahead and push changes for them.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/src/filesystem.cpp [hunk 13]
```diff
@@ -128,13 +128,14 @@ namespace {
...
     [[nodiscard]] __std_win_error __stdcall _Get_file_id_by_handle(
         const HANDLE _Handle, _Out_ FILE_ID_INFO* const _Id) noexcept {
-        __std_win_error _Last_error;
         if (GetFileInformationByHandleEx(_Handle, FileIdInfo, _Id, sizeof(*_Id)) != 0) {
             // if we could get FILE_ID_INFO, use that as the source of truth
             return __std_win_error::_Success;
         }
 
-        _Last_error = __std_win_error{GetLastError()};
+        __std_win_error _Last_error = __std_win_error{GetLastError()};
```
Comment (StephanTLavavej) [inline stl/src/filesystem.cpp:136]: As I mentioned before, we conventionally use direct-initialization here; `Type variable{initializer};` is simpler than `Type variable = Type{initializer};` with no loss of clarity. Precedent:

https://github.com/microsoft/STL/blob/9947dd9d6d2506029ed4486bb7596bda7d70f99d/stl/src/filesystem.cpp#L718

**Review by StephanTLavavej — APPROVED** [review]

File: stl/src/filesystem.cpp [hunk 14]
```diff
@@ -78,7 +78,7 @@ namespace {
         // take lower bits to undo HRESULT_FROM_WIN32
         return {false, __std_win_error{_Copy_result & 0x0000FFFFU}};
 #else // ^^^ defined(_CRT_APP) ^^^ // vvv !defined(_CRT_APP) vvv
-        if (CopyFileW(_Source, _Target, _Fail_if_exists)) {
+        if (CopyFileW(_Source, _Target, static_cast<BOOL>(_Fail_if_exists)) != 0) {
```
Comment (strega-nil-ms) [inline stl/src/filesystem.cpp:81]:
```suggestion
        if (CopyFileW(_Source, _Target, _Fail_if_exists)) {
```

this change is really weird and not good

**Comment by AtariDreams** [comment]
Isn't that the convention?

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@strega-nil-ms I pushed a merge with `main`, resolving a trivial merge conflict with #2802's detached comment. Then I audited `filesys.cpp` and `filesystem.cpp` for all comparisons of `BOOL` and `BOOLEAN` against `0` - both pre-existing occurrences, and ones that were introduced by this PR. (I checked every single function, instead of doing a search-and-replace, and was careful to not flip any comparisons accidentally.) I avoided changing the code that #2771 is replacing. I also reversed `DeviceIoControl()`'s flow to be if-success, since that's consistent with other changes here.

@AtariDreams Thanks for your patience - we should be able to merge this soon. The last thing is that we have to avoid significant changes to `msvcp140.dll` until VS 2022 17.4 Preview 2 is open for changes, so after Nicole double-checks the final changes here, we'll need to hold this PR in a `blocked` state for a little while along with a few other PRs we've accumulated.

File: stl/src/filesystem.cpp [hunk 15]
```diff
@@ -601,7 +596,7 @@ _Success_(return == __std_win_error::_Success) __std_win_error
             return {false, __std_win_error{GetLastError()}};
         }
         // check if FILE_ATTRIBUTE_READONLY is set
-        if (_Basic_info.FileAttributes & FILE_ATTRIBUTE_READONLY) {
+        if ((_Basic_info.FileAttributes & FILE_ATTRIBUTE_READONLY) != 0u) {
```
Comment (strega-nil-ms) [inline stl/src/filesystem.cpp:599]: I'm still curious why this is using `0u` instead of just `0`; no changes requested.

**Review by strega-nil-ms — APPROVED** [review]
Thanks, approved!

**Comment by strega-nil-ms** [comment]
This is now Ready To Merge, but I'm marking it as Blocked until changes start flowing into VS 2022 17.4 Preview 2 in a couple of weeks.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for finding and improving this filesystem code! :open_file_folder: :white_check_mark: :smile_cat:

**Pull request merged**